From 5d6b4845ca29dc6c9e426043f4c7ebf75257ec7f Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 22 Oct 2024 17:49:12 +0200
Subject: [PATCH 1449/1695] drivers: firewall: stm32_etzpc: consistent
 RCC/DECPROT secure config

Ensures that when an ETZPC DECPROT configuration is secure (resp.
MCU isolated) that SoC RCC is also secure (resp. MCKPROT isolated).

This change helps to remove dependency on shared_resource.c driver
that is no longer needed since integration of the firewall framework.

This change is back ported from commit 9c22da4b29de ("firewall:
stm32_etzpc: check consistency of RCC vs DECPROT secure config")

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 5a545889e903b661e94e049cec82092e8fd346c4 drivers: firewall: stm32_etzpc: consistent RCC/DECPROT secure config]
Change-Id: I3bfc7bc28e386274a819bee82385bf7c265b0bd8
Link: https://github.com/OP-TEE/optee_os/commit/9c22da4b29de0812f2dfc214107bba3c1c484ae8
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/413515
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/firewall/stm32_etzpc.c | 45 +++++++++++++++++++++++++++++
 1 file changed, 45 insertions(+)

diff --git a/core/drivers/firewall/stm32_etzpc.c b/core/drivers/firewall/stm32_etzpc.c
index 40c14ef5d..5acb1ffd9 100644
--- a/core/drivers/firewall/stm32_etzpc.c
+++ b/core/drivers/firewall/stm32_etzpc.c
@@ -13,6 +13,9 @@
 #include <assert.h>
 #include <drivers/clk_dt.h>
 #include <drivers/stm32_etzpc.h>
+#ifdef CFG_STM32MP15
+#include <drivers/stm32mp1_rcc.h>
+#endif
 #include <drivers/firewall.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <initcall.h>
@@ -166,6 +169,44 @@ static enum etzpc_decprot_attributes etzpc_binding2decprot(uint32_t mode)
 	}
 }
 
+static void
+sanitize_decprot_config(uint32_t decprot_id __maybe_unused,
+			enum etzpc_decprot_attributes attr __maybe_unused)
+{
+#ifdef CFG_STM32MP15
+	/*
+	 * STM32MP15: check dependency on RCC TZEN/MCKPROT configuration
+	 * when a ETZPC resource is secured or isolated for Cortex-M
+	 * coprocessor.
+	 */
+	switch (attr) {
+	case ETZPC_DECPROT_S_RW:
+	case ETZPC_DECPROT_NS_R_S_W:
+		if (!stm32_rcc_is_secure()) {
+			IMSG("WARNING: RCC tzen:0, insecure ETZPC hardening %"PRIu32":%s",
+			     decprot_id, etzpc_decprot_strings[attr]);
+			if (!IS_ENABLED(CFG_INSECURE))
+				panic();
+		}
+		break;
+	case ETZPC_DECPROT_MCU_ISOLATION:
+		if (!stm32_rcc_is_secure() || !stm32_rcc_is_mckprot()) {
+			IMSG("WARNING: RCC tzen:%u mckprot:%u, insecure ETZPC hardening %"PRIu32":%s",
+			     stm32_rcc_is_secure(), stm32_rcc_is_mckprot(),
+			     decprot_id, etzpc_decprot_strings[attr]);
+			if (!IS_ENABLED(CFG_INSECURE))
+				panic();
+		}
+		break;
+	case ETZPC_DECPROT_NS_RW:
+		break;
+	default:
+		assert(0);
+		break;
+	}
+#endif
+}
+
 static void etzpc_do_configure_decprot(struct etzpc_device *etzpc_dev,
 				       uint32_t decprot_id,
 				       enum etzpc_decprot_attributes attr)
@@ -181,6 +222,8 @@ static void etzpc_do_configure_decprot(struct etzpc_device *etzpc_dev,
 	FMSG("ID : %"PRIu32", CONF %s", decprot_id,
 	     etzpc_decprot_strings[attr]);
 
+	sanitize_decprot_config(decprot_id, attr);
+
 	exceptions = etzpc_lock(etzpc_dev);
 
 	io_clrsetbits32(base + ETZPC_DECPROT0 + offset,
@@ -534,6 +577,8 @@ static TEE_Result stm32_etzpc_configure_memory(struct firewall_query *firewall,
 				return TEE_ERROR_ACCESS_DENIED;
 			}
 
+			sanitize_decprot_config(id, attr);
+
 			return TEE_SUCCESS;
 		}
 
-- 
2.47.2

